src/node.cc

```diff
@@ -1110,6 +1111,55 @@ bool ShouldAbortOnUncaughtException(Isolate* isolate) {
 }
 
 
+void PromiseHook(PromiseHookType type,
+                 Local<Promise> promise,
+                 Local<Value> parent) {
+  Environment* env = Environment::GetCurrent(Isolate::GetCurrent());
+  Local<Context> context = env->context();
+
+  if (type == PromiseHookType::kInit && env->in_domain()) {
+    promise->Set(context,
+                 env->domain_string(),
+                 env->domain_array()->Get(0)).FromJust();
+  } else if (type == PromiseHookType::kBefore ||
+             type == PromiseHookType::kAfter) {
+    // Loosely based on node::MakeCallback().
+    Local<Value> domain_v =
+        promise->Get(context, env->domain_string()).ToLocalChecked();
+    if (!domain_v->IsObject())
+      return;
+
+    Local<Object> domain = domain_v.As<Object>();
+    if (domain->Get(context, env->disposed_string())
+            .ToLocalChecked()->IsTrue()) {
+      return;
+    }
+
+    if (type == PromiseHookType::kBefore) {
+      Local<Value> enter_v =
+          domain->Get(context, env->enter_string()).ToLocalChecked();
+      if (enter_v->IsFunction()) {
+        if (enter_v.As<Function>()->Call(context, domain, 0, nullptr)
+                .IsEmpty()) {
+          FatalError("node::PromiseHook",
+                     "domain enter callback threw, please report this");
+        }
+      }
+    } else {
+      Local<Value> exit_v =
+          domain->Get(context, env->exit_string()).ToLocalChecked();
+      if (exit_v->IsFunction()) {
+        if (exit_v.As<Function>()->Call(context, domain, 0, nullptr)
+                .IsEmpty()) {
```

Comment:
It’s odd to me too, I’ll gladly take suggestions. (Maybe just splitting the statement into multiple ones is the best way…)